README.md

```diff
@@ -20,13 +20,157 @@ The latest version is [![jcenter](https://api.bintray.com/packages/mercari-inc/m
 
 ``` Groovy
 repositories {
-    jcenter()
+  jcenter()
 }
 
 dependencies {
-    compile 'com.mercari.readerk:readerk:<latest-version>' 
+  compile 'com.mercari.readerk:readerk:<latest-version>' 
 }
 
 ```
 
-### Explanation
+## Explanation
+
+`Reader` helps us perform the same computation with different values.
+
+A simplest use-case for Reader monad is to use as DI (Dependency injection). Let's assume that we have a class that interacts with user account.
+
+```` Kotlin
+interface AccountService {
+  fun getAccount(accoundId: String, session: Session): Future<Account>
+  fun getBalance(account: Account, session: Session): Future<Amount>
+  fun getStatement(account: Account, session: Session): Future<Statement>
+}
+````
+
+And the concrete implementation like so
+
+```` Kotlin
+object Accounts : AccountService {
+
+  override fun getAccount(accoundId: String, session: Session): Future<Account> {
+    //real implementation
+  }
+  
+  override fun getBalance(account: Account, session: Session): Future<Amount> {
+    //real implementation
+  }
+  
+  override fun getStatement(account: Account, session: Session): Future<Statement> {
+    //real implementation
+  }
+}
+````
+
+It looks a bit annoying as we need to provide `session` for all of the functions. One can provide a DI at the constructor but it has 2 drawbacks.
+Firstly, your class becomes stateful and it is start to tangled with the construction of your dependency. Secondly, if you have multiple things to pass into constructor,
+the constructor becomes bloated and un-scalable.
+
+With Reader monad, you can improve above implementation to take advantage of `Reader`
+
+```` Kotlin
+typealias SessionReader<T> = Reader<Session, T>
+
+interface AccountService {
+  fun getAccount(accoundId: String): SessionReader<Future<Account>>
+  fun getBalance(account: Account): SessionsReader<Future<Amount>>
+  fun getStatement(account: Account): SessionReader<Future<Statement>>
+}
+
+object ReaderAccounts : AccountService {
+  
+  override fun getAccount(accoundId: String): SessionReader<Future<Account>> = 
+    Reader { session: Session -> 
+      session.doSomething() //do something before returning account, check authetication, status etc.
+      session.fetchAccount(accountId)
+    }
+  
+  override fun getBalance(account: Account): SessionsReader<Future<Amount>> =
+    Reader { session: Session ->
+      session.doSomething()
+      session.fetchAmount(account)
+    }
+  
+  override fun getStatement(account: Account): SessionReader<Future<Statement>> = 
+    Reader { session: Session ->
+      session.doSomething()
+      session.fetchStatement(account)
+    }
+}
+````
+
+As you can see now, we are using `Reader` to abstract away our dependency usage. `Reader` is meant to be composable so you can do something really cool like chaining a series of steps in one nice operation.
```

Comment:
At this line, the usual DI method `getSomething(account, session)` just got refactored into `getStatement(account): SessionReader<Future<Statement>>`, it will be good to see how we can use this on the call side right after it. 

but the next sentence talks about series chaining, so while I'm reading this, I am wondering how the call side looks like without the chaining, maybe it will be good to add it? 💭